### DISCUSSION TITLE
Observable Plot with NodeJS?

### DISCUSSION THREAD

**Discussion by nshiab** [observablehq/plot#847] [Q&A] [answered]
Hi!

I would love to use Plot on both front end and back end.

I tried using it with NodeJS, but it didn't work. Here's what I got.. I thought the function would output the svg code and I would be able to write it in a file. But it's looking for the html document and there isn't one.

Any idea how to make it work?

Thanks!
<img width="997" alt="Capture d’écran 2022-04-17 à 17 42 12" src="https://user-images.githubusercontent.com/11160343/163742094-7f3cd20c-e408-41ce-bf09-05bbc34e825b.png">

**Comment by mbostock** [comment, marked as answer]
You’ll need to provide a DOM implementation to use Plot. We use JSDOM for testing, so you could crib from our approach in [test/jsdom.js](https://github.com/observablehq/plot/blob/ff15257649f767cbd11d7d51f85945dbfd4eb36b/test/jsdom.js#L17-L41). For example, this script will output the generated HTML for a plot:

```js
import * as d3 from "d3";
import * as Plot from "@observablehq/plot";
import {JSDOM} from "jsdom";
import fs from "fs/promises";

withJsdom(async () => {
  const penguins = d3.csvParse(await fs.readFile("./penguins.csv", "utf8"), d3.autoType);
  const plot = Plot.plot({
    grid: true,
    marks: [
      Plot.dot(penguins, {
        x: "culmen_depth_mm",
        y: "culmen_length_mm",
        stroke: "species"
      })
    ]
  });
  console.log(plot.outerHTML);
})();

function withJsdom(run) {
  return async () => {
    const jsdom = new JSDOM("");
    global.document = jsdom.window.document;
    try {
      return await run();
    } finally {
      delete global.document;
    }
  };
}
```

I’ve also filed https://github.com/observablehq/plot/issues/848 so that you can pass _document_ as an option rather than requiring it to be a global.

**Reply by nshiab** [reply]
Wonderful! Thanks @mbostock !

**Reply by Fil** [reply]
I've put this solution into an easy-to install repo https://github.com/Fil/plot-cli
